### PR TITLE
moving PackUriHelper methods to S.IO.Packaging

### LINKED ISSUE #36459: missing methods on System.IO.Packaging.PackUriHelper that were in WindowsBase
This is the list of public types on the PackUriHelper class from WindowsBase that are no longer in the PackUriHelper as part of CoreFX:

```c#
        public static Uri GetPackageUri(Uri package);
        public static Uri Create(Uri uri);
        public static Uri Create(Uri uri, Uri partName);
        public static int ComparePackUri(Uri uri, Uri uri2);
```

These should be added back so applications porting to .NET Core aren't broken

### PR DESCRIPTION
**Pull request by stevenbrix**
The below methods are available in the PackUriHelper class exposed by WindowsBase. Now that this type is being removed and replaced (entirely) by the one here in S.IO.Packaging, we want to move some more functionality into this assembly.

Fixes #36459

### CODE REVIEW

**Comment by stevenbrix** [comment]
@ericstj it looks like some tests are building against netstandard1.3 and S.IO.Packaging is getting compiled as part of that? There are a few Uri API's as part of this PR that were added in netstandard2.0, so the build fails. What is the right course of action here?

**Comment by ericstj** [comment]
These aren’t tests but a build of the lib itself.  We will need to ifdef/partial class.  Maybe it’s easiest to just add new api only for ns2.0 and later?

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 1]
```diff
@@ -14,6 +14,59 @@ public static class PackUriHelper
...
+        /// <exception cref="ArgumentException">If packUri parameter is not an absolute Uri</exception>
+        /// <exception cref="ArgumentException">If packUri parameter does not have "pack://" scheme</exception>
+        /// <exception cref="ArgumentException">If inner packageUri extracted from the packUri has a fragment component</exception>
+        public static Uri GetPackageUri(Uri packUri)
+        {
+            Uri packageUri;
+            Uri partUri;
+
+            //Parameter Validation is done in the follwoing method
+            ValidateAndGetPackUriComponents(packUri, out packageUri, out partUri);
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Nit: this could just use a discard for partUri, since it's never used.

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 2]
```diff
@@ -295,6 +348,55 @@ public static Uri GetSourcePartUriFromRelationshipPartUri(Uri relationshipPartUr
...
+        /// Value - Equal to zero means both the Uris are equal
+        /// Value - Greater than zero means firstUri is greater than secondUri </returns>
+        /// <exception cref="ArgumentException">If either of the Uris are not absolute or if either of the Uris are not with pack:// scheme</exception>
+        /// <exception cref="ArgumentException">If firstPackUri or secondPackUri parameter is not an absolute Uri</exception>
+        /// <exception cref="ArgumentException">If firstPackUri or secondPackUri parameter does not have "pack://" scheme</exception>
+        public static int ComparePackUri(Uri firstPackUri, Uri secondPackUri)
+        {
+            //If any of the operands are null then we simply call System.Uri compare to return the correct value
+            if (firstPackUri == null || secondPackUri == null)
+                return CompareUsingSystemUri(firstPackUri, secondPackUri);
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Nit: please use braces around this block

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 3]
```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
...
+            // Step 1 - Validate input parameters
+            packageUri = ValidatePackageUri(packageUri);
+
+            if (partUri != null)
+                partUri = ValidatePartUri(partUri);
+
+            if (fragment != null)
+            {
+                if (fragment == String.Empty || fragment[0] != '#')
+                    throw new ArgumentException(SR.FragmentMustStartWithHash);
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Nit: could we add a parameter name to this?

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 4]
```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
...
+        /// This parameter can be null in which case we will create a valid pack uri
+        /// that references the entire package</param>
+        /// <param name="fragment">Fragment for the resulting Pack URI. This parameter can be null
+        /// The fragment string must start with a "#"</param>
+        /// <returns>A Uri with the "pack://" scheme</returns>
+        /// <exception cref="ArgumentNullException">If packageUri parameter is null</exception>
+        /// <exception cref="ArgumentException">If packageUri parameter is not an absolute Uri</exception>
+        /// <exception cref="ArgumentException">If partUri parameter does not conform to the valid partUri syntax</exception>
+        /// <exception cref="ArgumentException">If fragment parameter is empty or does not start with a "#"</exception>
+        private static Uri Create(Uri packageUri, Uri partUri, string fragment)
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Maybe I'm missing it, but is this ever called with a non-null fragment? If no, it would seem we could just assert that and then delete the dead code below for the non-null case.

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 5]
```diff
@@ -295,6 +348,55 @@ public static Uri GetSourcePartUriFromRelationshipPartUri(Uri relationshipPartUr
...
+
+                Uri firstPackageUri;
+                Uri secondPackageUri;
+                Uri firstPartUri;
+                Uri secondPartUri;
+
+                ValidateAndGetPackUriComponents(firstPackUri, out firstPackageUri, out firstPartUri);
+                ValidateAndGetPackUriComponents(secondPackUri, out secondPackageUri, out secondPartUri);
+
+
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Nit: unnecessary extra blank line

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 6]
```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
...
+        /// <param name="packageUri"></param>
+        /// <returns></returns>
+        private static Uri ValidatePackageUri(Uri packageUri)
+        {
+            if (packageUri == null)
+                throw new ArgumentNullException(nameof(packageUri));
+
+            if (!packageUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Not: can we add a parameter name?

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 7]
```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
...
+        private static Uri ValidatePackUri(Uri packUri)
+        {
+            if (packUri == null)
+                throw new ArgumentNullException(nameof(packUri));
+
+            if (!packUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            if (packUri.Scheme != PackUriHelper.UriSchemePack)
+                throw new ArgumentException(SR.UriShouldBePackScheme);
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: Nit: I'll stop commenting on these, but it'd be great to audit all of these arg exceptions to make sure they contain names whenever relevant.

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs [hunk 8]
```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
...
+        {            
+            string characterString;
+
+            // Escaping for the following - '%'; '@'; ',' and '?'
+            // !!Important!! - The order is important - The '%' sign should be escaped first.
+            // This is currently enforced by the order of characters in the _specialCharacters array
+            foreach (char c in s_specialCharacterChars)
+            {
+                characterString = c.ToString();
+                if (path.Contains(characterString))
```
Comment (stephentoub) [inline src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: This ToString should not be necessary, at least not outside of the if block. String.Contains has an overload that takes a char.

Comment (stephentoub) [reply src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: And it really shouldn't ever be necessary, since this is a fixed list of chars for which we could just store the strings.

File: src/System.IO.Packaging/tests/Tests.cs [hunk 9]
```diff
@@ -3774,6 +3774,93 @@ public void GetPartCallsGetPartCore()
...
+        public void ComparePackUriSamePackSamePart()
+        {
+            Uri partUri = new Uri("/idontexist.xml", UriKind.Relative);
+            Uri packageUri = new Uri("application://");
+
+            Uri combinedUri = PackUriHelper.Create(packageUri, partUri);
+            Assert.Equal(combinedUri.ToString(), "pack://application:,,,/idontexist.xml");
+
+            Uri sameCombinedUri = PackUriHelper.Create(packageUri, partUri);
+            Assert.Equal(combinedUri.ToString(), "pack://application:,,,/idontexist.xml");
```
Comment (stephentoub) [inline src/System.IO.Packaging/tests/Tests.cs]: Nit: The arguments here and to some of the other asserts should be reversed. The first argument is "expected", the second is "actual".

File: src/System.IO.Packaging/tests/Tests.cs [hunk 10]
```diff
@@ -3774,6 +3774,93 @@ public void GetPartCallsGetPartCore()
...
+
+            Uri returnedPackageUri = PackUriHelper.GetPackageUri(combinedUri);
+            Uri returnedSamePackageUri = PackUriHelper.GetPackageUri(sameCombinedUri);
+
+            // Validate the PackageUri returned from PackUriHelper.GetPackageHelper matches what was given to PackUriHelper.Create
+            Assert.Equal(packageUri, returnedPackageUri);
+            Assert.Equal(packageUri, returnedSamePackageUri);
+
+            // Validate PackUriHelper.ComparePackUri correctly validates identical pack uri's.
+            Assert.True(PackUriHelper.ComparePackUri(combinedUri, sameCombinedUri) == 0);
```
Comment (stephentoub) [inline src/System.IO.Packaging/tests/Tests.cs]: Assert.Equal

File: src/System.IO.Packaging/tests/Tests.cs [hunk 11]
```diff
@@ -3774,6 +3774,93 @@ public void GetPartCallsGetPartCore()
...
+            // Validate the PackageUri returned from PackUriHelper.GetPackageHelper matches what was given to PackUriHelper.Create
+            Assert.Equal(packageUri, returnedPackageUri);
+            Assert.Equal(packageUri, returnedPackageUriNoPart);
+            Assert.Equal(packageUri, returnedPackageUriDifferentPart);
+
+            // Validate PackUriHelper.ComparePackUri correctly compares pack uri's with different parts. These are not
+            // considered equal because the parts are different.
+            Assert.False(PackUriHelper.ComparePackUri(combinedUriWithPart, combinedUriWithDifferentPart) == 0);
+            Assert.False(PackUriHelper.ComparePackUri(combinedUriWithPart, combinedUriNoPart) == 0);
+            Assert.False(PackUriHelper.ComparePackUri(combinedUriNoPart, combinedUriWithDifferentPart) == 0);
```
Comment (stephentoub) [inline src/System.IO.Packaging/tests/Tests.cs]: Assert.NotEqual

File: src/System.IO.Packaging/tests/Tests.cs [hunk 12]
```diff
@@ -3774,6 +3774,93 @@ public void GetPartCallsGetPartCore()
...
+            Uri returnedDifferentPackageUri = PackUriHelper.GetPackageUri(differentPackageSamePart);
+
+            // Validate the PackageUri returned from PackUriHelper.GetPackageHelper matches what was given to PackUriHelper.Create
+            Assert.Equal(packageUri, returnedPackageUri);
+            Assert.Equal(packageUri, returnedSamePackageUri);
+            Assert.Equal(differentPackageUri, returnedDifferentPackageUri);
+
+            // Validate PackUriHelper.ComparePackUri correctly compares pack uri's with different packages.
+            Assert.False(PackUriHelper.ComparePackUri(packageUriWithPart, differentPackageSamePart) == 0);
+            Assert.False(PackUriHelper.ComparePackUri(samePackageNoPart, differentPackageSamePart) == 0);
```
Comment (stephentoub) [inline src/System.IO.Packaging/tests/Tests.cs]: Assert.NotEqual

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs (hunk 3 above)
Comment (stevenbrix) [reply src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: the parameter name was baked into the resource string, but i'll bring it out so i can use SR.Format.

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs (hunk 4 above)
Comment (ericstj) [reply src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: I had it in here https://github.com/dotnet/corefx/issues/29959, but it wasn't hit when you ripped the code out of WindowsBase, so my guess is that it wasn't used by WindowsBase?  I'm OK adding it back as public, we shouldn't subset the public API unless we have a good reason.

**Comment by ericstj** [comment]
We can't remove the netstandard1.3 config.  Instead can you put all your new code in a new file that is a partial class?  I can help with this if you like.

File: src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs (hunk 4 above)
Comment (stevenbrix) [reply src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs]: sounds good, i'll make it public and add a test for it

**Comment by stevenbrix** [comment]
> We can't remove the netstandard1.3 config.  Instead can you put all your new code in a new file that is a partial class?  I can help with this if you like.

thanks eric, i'll let you know if i need any help with this

File: src/System.IO.Packaging/ref/Configurations.props [hunk 13]
```diff
@@ -1,7 +1,6 @@
 ﻿<Project>
   <PropertyGroup>
     <PackageConfigurations>
-      netstandard1.3;
```
Comment (stevenbrix) [inline src/System.IO.Packaging/ref/Configurations.props]: will undo these changes here
